### ISSUE TITLE
Revisions to deformdemo on master

### ISSUE THREAD

**Issue by stevepiercy** [Pylons/deformdemo#87]
@sydoluciani FYI I made a few changes to deformdemo on master because it should run all versions of Python and I found a simpler way to instantiate `browser`. Please make sure you pull `master` and compare the differences. Ignore all the fiddling around with lint, as I am juggling two things at once and did not want to interrupt the docker container testing by updating to the latest version of black.

**Comment by stevepiercy** [comment]
Here's a diff:
https://github.com/Pylons/deformdemo/compare/3e2ed6e5e2a4d09660991fc94115c9e919ac5239...master

Please let me know if you have any questions.

**Comment by sydoluciani** [comment]
@stevepiercy yes, that is a good reduction, if firefox is in path, as we explained it in contributing.md, then there will never need for specifying FIREFOX_PATH.

Removing [functional3](https://github.com/Pylons/deformdemo/compare/3e2ed6e5e2a4d09660991fc94115c9e919ac5239...master#diff-b91f3d5bd63fcd17221b267e851608e8L26) causes [commands](https://github.com/Pylons/deformdemo/blob/master/tox.ini#L27) listed in testenv section run in every other sub section of testenv, and that includes lint. so [functional tests](https://github.com/Pylons/deformdemo/blob/master/tox.ini#L27) are actually being run but being run part of lint. I think it is better to have it's own section. if you actually run tox within Deformdemo, it starts pserve and run all tests against server, but it is running it part of lint. for now there is only two subsection, lint and format, but still it is better to have its own subsection.

**Comment by sydoluciani** [comment]
@stevepiercy If I am right, then running tox in Deformdemo causes tests run two times, once in lint and once in format. Please let me know if this is not the case.

**Comment by stevepiercy** [comment]
@sydoluciani correct, I was sloppy as a compromise to get all versions to run tests locally and in Travis. I'm working on fixing that up next.

**Comment by stevepiercy** [comment]
To give further explanation, it appeared that tox was not actually running the functional tests in Travis. Take a look:

- https://travis-ci.org/github/Pylons/deformdemo/builds/723664180
- https://travis-ci.org/github/Pylons/deformdemo/builds/724606426

Here's the full deformdemo build history: https://travis-ci.org/github/Pylons/deformdemo/builds

Each job took about 1 minute, instead of the usual 5 or 6 minutes, for running functional tests. Additionally there was no output in the report other than "congratulations", compared to the more informative feedback with the previous method.

Compare your py38 environment to mine:
- https://travis-ci.org/github/Pylons/deformdemo/jobs/724606430#L241-L244
- https://travis-ci.org/github/Pylons/deformdemo/jobs/725107933#L242-L343

My interpretation of these results was that the functional tests were not actually being run on Travis. That could be an incorrect interpretation. Please let me know.

**Comment by sydoluciani** [comment]
> To give further explanation, it appeared that tox was not actually running the functional tests in Travis. Take a look:
> 
> * https://travis-ci.org/github/Pylons/deformdemo/builds/723664180
> * https://travis-ci.org/github/Pylons/deformdemo/builds/724606426
> 
> Here's the full deformdemo build history: https://travis-ci.org/github/Pylons/deformdemo/builds
> 
> Each job took about 1 minute, instead of the usual 5 or 6 minutes, for running functional tests. Additionally there was no output in the report other than "congratulations", compared to the more informative feedback with the previous method.
> 
> Compare your py38 environment to mine:
> 
> * https://travis-ci.org/github/Pylons/deformdemo/jobs/724606430#L241-L244
> * https://travis-ci.org/github/Pylons/deformdemo/jobs/725107933#L242-L343
> 
> My interpretation of these results was that the functional tests were not actually being run on Travis. That could be an incorrect interpretation. Please let me know.

Yes, problem started after I added functional3 environment but forgot to add it to .travis.yaml.
If you could please make it the same as Deform.

we should be able to run the same command in Deform and Deformdemo to start functional3.
currently we can't run tox -e functional3 within Defromdemo, since functional3 envrionment does not exist.
We should introduce functional3 to both tox.ini and .travis.yaml files in Deformdemo to be the same as Deform.

**Comment by stevepiercy** [comment]
Adding functional3 to `.travis.yml` has no impact to running tox locally. Sorry, I don't see the advantage of locally running functional3 the way you had it configured in tox.

```
git checkout 3a6bd63e9c6c9673d8a444bc5b9f3cfe9092d4cc
$TOX
```

That has a similar result to that in Travis. Here are just two successive environment output, each of which takes a few seconds to run, not the expected 5 minutes or so.

```
py35 recreate: /Users/stevepiercy/projects/deform/deformdemo_functional_tests/.tox/py35
py35 develop-inst: /Users/stevepiercy/projects/deform/deformdemo_functional_tests
py35 installed: Babel==2.8.0,Chameleon==3.8.1,colander==1.8.2,deform==2.0.14,-e git+git@github.com:stevepiercy/deformdemo.git@3a6bd63e9c6c9673d8a444bc5b9f3cfe9092d4cc#egg=deformdemo,hupper==1.10.2,iso8601==0.1.12,PasteDeploy==2.1.0,peppercorn==0.6,plaster==1.0,plaster-pastedeploy==0.7,Pygments==2.6.1,pyramid==1.10.4,pyramid-chameleon==0.3,pytz==2020.1,six==1.15.0,translationstring==1.4,venusian==3.0.0,waitress==1.4.4,WebOb==1.8.6,zope.deprecation==4.4.0,zope.interface==5.1.0
py35 run-test-pre: PYTHONHASHSEED='1155213043'
py36 recreate: /Users/stevepiercy/projects/deform/deformdemo_functional_tests/.tox/py36
py36 develop-inst: /Users/stevepiercy/projects/deform/deformdemo_functional_tests
py36 installed: Babel==2.8.0,Chameleon==3.8.1,colander==1.8.2,deform==2.0.14,-e git+git@github.com:stevepiercy/deformdemo.git@3a6bd63e9c6c9673d8a444bc5b9f3cfe9092d4cc#egg=deformdemo,hupper==1.10.2,iso8601==0.1.12,PasteDeploy==2.1.0,peppercorn==0.6,plaster==1.0,plaster-pastedeploy==0.7,Pygments==2.6.1,pyramid==1.10.4,pyramid-chameleon==0.3,pytz==2020.1,six==1.15.0,translationstring==1.4,venusian==3.0.0,waitress==1.4.4,WebOb==1.8.6,zope.deprecation==4.4.0,zope.interface==5.1.0
py36 run-test-pre: PYTHONHASHSEED='1155213043'
```

and ultimately:

```
functional3 create: /Users/stevepiercy/projects/deform/deformdemo_functional_tests/.tox/functional3
ERROR: InterpreterNotFound: 
```

`$TOX` should run all environments specified in `envlist` under `[tox]` in both Deform and deformdemo. Deform has a special `functional3` environment that runs only under one version of Python, and that is OK.

**Comment by sydoluciani** [comment]
Not sure how it is working but you are right, functional tests are only run under py3* environments and not under lint or format, so no need to add functional3.

**Comment by stevepiercy** [comment]
OK, thanks for confirming.

**Comment by stevepiercy** [comment]
Closed by #88
